features/remotepollerng/daemon/src/main/java/org/opennms/netmgt/remotepollerng/PerspectiveServiceTracker.java

```diff
@@ -0,0 +1,259 @@
+        private final String serviceName;
+        private final String perspectiveLocation;
+
+        public ServicePerspectiveRef(final int nodeId,
+                                     final InetAddress ipAddress,
+                                     final String serviceName,
+                                     final String perspectiveLocation) {
+            this.nodeId = Objects.requireNonNull(nodeId);
+            this.ipAddress = Objects.requireNonNull(ipAddress);
+            this.serviceName = Objects.requireNonNull(serviceName);
+            this.perspectiveLocation = Objects.requireNonNull(perspectiveLocation);
+        }
+
+        public int getNodeId() {
+            return this.nodeId;
+        }
+
+        public InetAddress getIpAddress() {
+            return this.ipAddress;
+        }
+
+        public String getServiceName() {
+            return this.serviceName;
+        }
+
+        public String getPerspectiveLocation() {
+            return this.perspectiveLocation;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) {
+                return true;
+            }
+            if (!(o instanceof ServicePerspectiveRef)) {
+                return false;
+            }
+            final ServicePerspectiveRef that = (ServicePerspectiveRef) o;
+            return Objects.equals(this.nodeId, that.nodeId) &&
+                   Objects.equals(this.ipAddress, that.ipAddress) &&
+                   Objects.equals(this.serviceName, that.serviceName) &&
+                   Objects.equals(this.perspectiveLocation, that.perspectiveLocation);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(this.nodeId, this.ipAddress, this.serviceName, this.perspectiveLocation);
+        }
+
+        public static ServicePerspectiveRef from(final ServicePerspective servicePerspective) {
+            return new ServicePerspectiveRef(servicePerspective.getService().getNodeId(),
+                                             servicePerspective.getService().getIpAddress(),
+                                             servicePerspective.getService().getServiceName(),
+                                             servicePerspective.getPerspectiveLocation().getLocationName());
+        }
+    }
+
+    private final SessionUtils sessionUtils;
+
+    private final ApplicationDao applicationDao;
+
+    private final Set<Session> sessions = Sets.newHashSet();
+
+    private final Timer timer = new Timer("PerspectiveServiceTracker-Timer");
+
+    @Autowired
+    public PerspectiveServiceTracker(final SessionUtils sessionUtils,
+                                     final ApplicationDao applicationDao) {
+        this.sessionUtils = Objects.requireNonNull(sessionUtils);
+        this.applicationDao = Objects.requireNonNull(applicationDao);
+
+        final long timerIntevalMs = Math.min(REFRESH_RATE_LIMIT_MS, TimeUnit.SECONDS.toMillis(5));
```

Comment:
Do we intend to use `min` (vs `max` here?). The default value of `REFRESH_RATE_LIMIT_MS` is 30 seconds.

Reply:
I think so. This is the polling interval denoting how often it checks for outstanding refresh (those who are not executed immediately due to the rate limit). This should usually be smaller than the rate limit.

Reply:
Ah yes, got it. Every 5ms the timer checks if an update is necessary and updates if the last one is older than REFRESH_RATE_LIMIT_MS, right? This way events are handled faster than waiting for REFRESH_RATE_LIMIT_MS.

Reply:
Exactly (beside the check is every 5s not every 5ms)

Reply:
Right.